godot-core/src/builtin/callable.rs

```diff
@@ -0,0 +1,190 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use godot_ffi as sys;
+
+use crate::builtin::{inner, ToVariant, Variant};
+use crate::engine::Object;
+use crate::obj::{Gd, GodotClass, InstanceId};
+use std::fmt;
+use sys::{ffi_methods, GodotFfi};
+
+use super::{StringName, VariantArray};
+
+/// A `Callable` represents a function in Godot.
+///
+/// Usually a callable is a reference to an `Object` and a method name, this is a standard callable. See
+/// [`Callable::is_standard`]. But can also be a custom callable, which is usually created from `bind`,
+/// `unbind`, or a GDScript lambda. See [`Callable::is_custom`].
+///
+/// Currently it is impossible to use `bind` and `unbind` in GDExtension, see [godot-cpp#802].
+///
+/// [godot-cpp#802]: https://github.com/godotengine/godot-cpp/issues/802
+#[repr(C, align(8))]
+pub struct Callable {
+    opaque: sys::types::OpaqueCallable,
+}
+
+impl Callable {
+    fn from_opaque(opaque: sys::types::OpaqueCallable) -> Self {
+        Self { opaque }
+    }
+
+    /// Create a callable for the method named `method` in the specified `object`.
+    ///
+    /// _Godot equivalent: `Callable(Object object, StringName method)`_
+    pub fn from_object_method<T, S>(object: Gd<T>, method: S) -> Self
+    where
+        T: GodotClass, // + Inherits<Object>,
+        S: Into<StringName>,
+    {
+        // upcast not needed
+        let method = method.into();
+        unsafe {
+            Self::from_sys_init_default(|self_ptr| {
+                let ctor = sys::builtin_fn!(callable_from_object_method);
+                let args = [object.sys_const(), method.sys_const()];
+                ctor(self_ptr, args.as_ptr());
+            })
+        }
+    }
+
+    /// Calls the method represented by this callable.
+    ///
+    /// Arguments passed should match the method's signature. If called with more arguments than expected by
+    /// the method, the extra arguments will be ignored. If called with fewer arguments than expected it will
+    /// crash. If called with arguments of the wrong type then an error will be printed and the call fails.
+    ///
+    /// _Godot equivalent: `callv`_
+    pub fn callv(&self, arguments: VariantArray) -> Variant {
+        self.as_inner().callv(arguments)
+    }
```

Comment:
Error conditions are a bit vague:
* "it will crash" -- UB or what?
* "the call fails" -- panic? NIL returned?

It's also not clear what happens when the callable is null (empty) or invalid (bad method name).

Reply:
I'll check

Reply:
Could maybe be worth adding as tests too. These things tend to change in Godot patch versions 😁

Reply:
the ones that can be tested without crashing our integration tests entirely are tested now. also we can't check that an error is printed in godot.

Reply:
That's more than enough, thanks a lot! 👍